src/magic_folder/util/database.py

```diff
@@ -0,0 +1,132 @@
+# See COPYING for details.
+
+"""
+Utilties for dealing with sqlite.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import contextlib
+import inspect
+import sqlite3
+
+import attr
+from six import wraps
+from twisted.python.compat import currentframe
+
+__all__ = [
+    "LockableDatabase",
+    "RecusiveTransaction",
+    "WithCursorGenerator",
+    "with_cursor",
+]
+
+
+@attr.s(auto_exc=True, frozen=True)
+class WithCursorGenerator(TypeError):
+    """
+    :py:`with_cursor` cannot be used on a generator.
+
+    As it is currently implemented, the transaction would be commited before
+    any code in the decorated generator ran. Even if we special cased generators,
+    having the transaction stay open until the generator completes, that loses
+    the proprerty of `with_cursor` being transparent and atomic from the point
+    of view of the calller.
+    """
+
+    function = attr.ib()
+
+    def __str__(self):
+        return "'with_cursor' cannot decorate a generator function: {!r}".format(
+            self.function
+        )
+
+
+@attr.s(auto_exc=True, frozen=True)
+class RecusiveTransaction(Exception):
+    function = attr.ib(default=None)
+    caller = attr.ib(default=None)
+
+    def __str__(self):
+        message = "Tried to enter a recursive transaction"
+        if self.function:
+            message += " when calling '{}'".format(self.function)
+        if self.caller:
+            message += " from '{}'".format(self.caller)
+        message += "."
+        return message
+
+
+@attr.s
+class LockableDatabase(object):
+    """
+    Wrapper around an :py:`sqlite3.Connection`
+    """
+
+    _database = attr.ib(validator=attr.validators.instance_of(sqlite3.Connection))
+    _in_transaction = attr.ib(
+        default=False, init=False, validator=attr.validators.instance_of(bool)
+    )
+
+    @contextlib.contextmanager
+    def transaction(self, function_name=None):
+        """
+        Context manager that starts a transaction and rollsback
+        if an exception is raised.
+
+        :param Optional[function_name]: The name of the calling function, used
+            in error messages.
+
+        :raises RecusiveTransaction: if used recusively.
+
+        :return sqlite3.Cursor:
+        """
+        if self._in_transaction:
+            frame = currentframe(2)
+            if frame.f_code == with_cursor._wrapped_code:
+                frame = frame.f_back
+            caller = frame.f_code.co_name
+            raise RecusiveTransaction(function=function_name, caller=caller)
+        try:
+            self._in_transaction = True
+            with self._database:
+                cursor = self._database.cursor()
+                cursor.execute("BEGIN IMMEDIATE TRANSACTION")
+                yield cursor
+        finally:
+            self._in_transaction = False
+
+
+# XXX: with_cursor lacks unit tests, see:
+#      https://github.com/LeastAuthority/magic-folder/issues/173
```

Comment:
So close to being fully tested now!  But I guess that just means #173 will be easily resolved with a follow-up PR.